File: src/main/java/eu/virtusdevelops/easyclans/dao/DaoCrud.java
package eu.virtusdevelops.easyclans.dao;

import java.util.List;

public interface DaoCrud<T, ID>{
    /**
     * Used for initializing database or required stuff
     */
    void init();

    /**
     * Gets item by id from database
     * @param id
     * @return
     */
    T getById(ID id);
    List<T> getAll();

    /**
     * Inserts new or updates if it already exists
     * @param t data to save
     * @return saved model or null if failure
     */
    T save(T t);
    boolean delete(T t);
}

File: core/src/main/kotlin/eu/virtusdevelops/easyclans/core/database/DaoCrud.kt
package eu.virtusdevelops.easyclans.core.database

import eu.virtusdevelops.easyclans.api.Success

interface DaoCrud <T, ID> {

    suspend fun init(): Result<Success>

    suspend fun getById(id: ID): Result<T>

    suspend fun getAll(): Result<List<T>>

    suspend fun save(t: T): Result<Success>

    suspend fun deleteById(id: ID): Result<Success>
}